In [1]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt
import tensorflow as tf
from PIL import Image
import os
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from keras.utils import to_categorical
from keras.models import Sequential, load_model
from keras.layers import Conv2D, MaxPool2D, Dense, Flatten, Dropout

Using TensorFlow backend.


In [ ]:
# !pip freeze>requirements.txt

In [2]:
data = []
labels = []
classes = 43 # predefined
cur_path = os.getcwd()
# print(cur_path)

In [3]:
#Getting images and labels 
for i in range(classes):
    path = os.path.join(cur_path,'Train',str(i)) #Each class folder is it's class number
    images = os.listdir(path)

    for img in images:
        try:
            image = Image.open(path + '\\'+ img)
            image = image.resize((30,30))
            image = np.array(image)
            data.append(image)
            labels.append(i)
        except:
            print("Error loading image")

In [4]:
#Converting data/label into numpy arrays
data = np.array(data)
labels = np.array(labels)

In [9]:
print(data.shape, labels.shape)

(39209, 30, 30, 3) (39209,)


Number of images: 39209 <br>
Size of images: 30x30<br>
Channels in each image: 3 (RGB)

In [5]:
#80-20 split(train and test) of dataset
X_train, X_test, y_train, y_test = train_test_split(data, labels, test_size=0.2, random_state=42)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(31367, 30, 30, 3) (7842, 30, 30, 3) (31367,) (7842,)


In [6]:
#One hot encoding of labels
y_train = to_categorical(y_train, 43) #43 classes
y_test = to_categorical(y_test, 43)

In [7]:
#Building the model
model = Sequential()
model.add(Conv2D(filters=32, kernel_size=(5,5), activation='relu', input_shape=X_train.shape[1:]))
model.add(Conv2D(filters=32, kernel_size=(5,5), activation='relu'))
model.add(MaxPool2D(pool_size=(2, 2)))
model.add(Dropout(rate=0.10))
model.add(Conv2D(filters=64, kernel_size=(3, 3), activation='relu'))
model.add(Conv2D(filters=64, kernel_size=(3, 3), activation='relu'))
model.add(MaxPool2D(pool_size=(2, 2)))
model.add(Dropout(rate=0.5))
model.add(Flatten())
model.add(Dense(256, activation='relu'))
model.add(Dropout(rate=0.5))
model.add(Dense(43, activation='softmax'))

In [8]:
#Compilation of the model
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

In [15]:
epochs = 10
model.fit(X_train, y_train, batch_size=64, epochs=epochs, validation_data=(X_test, y_test))

Epoch 1/10
491/491 [==============================] - 79s 162ms/step - loss: 0.2980 - accuracy: 0.9104 - val_loss: 0.0844 - val_accuracy: 0.9768
Epoch 2/10
491/491 [==============================] - 97s 197ms/step - loss: 0.2941 - accuracy: 0.9119 - val_loss: 0.0805 - val_accuracy: 0.9786
Epoch 3/10
491/491 [==============================] - 99s 201ms/step - loss: 0.2768 - accuracy: 0.9186 - val_loss: 0.0810 - val_accuracy: 0.9772
Epoch 4/10
491/491 [==============================] - 93s 190ms/step - loss: 0.2682 - accuracy: 0.9199 - val_loss: 0.0709 - val_accuracy: 0.9802
Epoch 5/10
491/491 [==============================] - 94s 191ms/step - loss: 0.2655 - accuracy: 0.9226 - val_loss: 0.0795 - val_accuracy: 0.9774
Epoch 6/10
491/491 [==============================] - 88s 178ms/step - loss: 0.2536 - accuracy: 0.9259 - val_loss: 0.0775 - val_accuracy: 0.9791
Epoch 7/10
491/491 [==============================] - 73s 148ms/step - loss: 0.2290 - accuracy: 0.9322 - val_loss: 0.0637 - val_ac

In [14]:
# plt.figure(0)
# plt.plot(history.history['accuracy'], label='train accuracy')
# plt.plot(history.history['val_accuracy'], label='val accuracy')
# plt.title('Accuracy')
# plt.xlabel('epochs')
# plt.ylabel('accuracy')
# plt.legend()
# plt.show()

In [18]:
model.save("traffic_sign_detector_model.h5")
# pickle.dump(model,open('traffic_sign_detector_model.pkl', 'wb'))

In [19]:
#testing accuracy on test dataset
y_test = pd.read_csv('Test.csv') #csv file contains test images path and class labels

labels = y_test["ClassId"].values
images = y_test["Path"].values

data=[]

for img in imgs:
    image = Image.open(img)
    image = image.resize((30,30))
    data.append(np.array(image))

X_test=np.array(data)

In [21]:
# model=pickle.load(open('traffic_sign_detector_model.pkl', 'rb'))
model=load_model("traffic_sign_detector_model.h5")

In [22]:
pred = model.predict_classes(X_test)

Instructions for updating:
Please use instead:* `np.argmax(model.predict(x), axis=-1)`,   if your model does multi-class classification   (e.g. if it uses a `softmax` last-layer activation).* `(model.predict(x) > 0.5).astype("int32")`,   if your model does binary classification   (e.g. if it uses a `sigmoid` last-layer activation).


In [23]:
#Comparing with ground truth to get accuracy
print(accuracy_score(labels, pred))

0.948535233570863


In [40]:
y_test = pd.read_csv('Test.csv')
images = y_test["Path"].values
file_path=images[382]
# file_path="warning.jpg"
image = Image.open(file_path)
image = image.resize((30,30))
image = np.expand_dims(image, axis=0)
image = np.array(image)
pred = model.predict_classes([image])[0]
sign = pred
print(sign)

14
